clippy_lints/src/methods/case_sensitive_file_extension_comparisons.rs

```diff
@@ -28,13 +38,36 @@ pub(super) fn check<'tcx>(
         let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
         if recv_ty.is_str() || is_type_lang_item(cx, recv_ty, LangItem::String);
         then {
-            span_lint_and_help(
+            span_lint_and_then(
                 cx,
                 CASE_SENSITIVE_FILE_EXTENSION_COMPARISONS,
-                call_span,
+                recv.span.to(call_span),
                 "case-sensitive file extension comparison",
-                None,
-                "consider using a case-insensitive comparison instead",
+                |diag| {
+                    diag.help("consider using a case-insensitive comparison instead");
+                    let mut recv_source = Sugg::hir(cx, recv, "").to_string();
+
+                    if is_type_lang_item(cx, recv_ty, LangItem::String) {
```

Comment:
`recv_ty` has had its references peeled, but if it's a `&String` already we don't need to add a `&`, so this could be

```suggestion
                    if !cx.typeck_results().expr_ty(recv).is_ref() {
```

(or change `recv_ty` to not use `.peel_refs()` and peel it separately for line 39)